src/outputs/alsa.c

```diff
@@ -1185,6 +1185,99 @@ alsa_write(struct output_buffer *obuf)
     }
 }
 
+static void
+alsa_device_list(const char *hwdev, int card, snd_ctl_card_info_t *info, snd_mixer_selem_id_t *sid)
+{
+  snd_ctl_t *hdl = NULL;
+  snd_mixer_t *mixer = NULL;
+  snd_mixer_elem_t *elem;
+  bool has_playback = false;
+
+  int err;
+
+  err = snd_ctl_open(&hdl, hwdev, 0);
+  if (err < 0)
+    {
+      DPRINTF(E_WARN, L_LAUDIO, "Failed to probe ALSA card=%d - %s\n", card, snd_strerror(err));
+      return;
+    }
+
+  err = snd_ctl_card_info(hdl, info);
+  if (err < 0)
+    {
+      DPRINTF(E_WARN, L_LAUDIO, "Failed to probe ALSA (info) card=%d - %s\n", card, snd_strerror(err));
+      goto error;
+    }
+
+  err = snd_mixer_open(&mixer, 0);
+  if (err < 0)
+    {
+      DPRINTF(E_WARN, L_LAUDIO, "Failed to probe ALSA (mixer open) card=%d - %s\n", card, snd_strerror(err));
+      goto error;
+    }
+
+  err = snd_mixer_attach(mixer, hwdev);
+  if (err < 0)
+    {
+      DPRINTF(E_WARN, L_LAUDIO, "Failed to probe ALSA (mixer attach) card=%d - %s\n", card, snd_strerror(err));
+      goto errormixer;
+    }
+
+  err = snd_mixer_selem_register(mixer, NULL, NULL);
+  if (err < 0)
+    {
+      DPRINTF(E_WARN, L_LAUDIO, "Failed to probe ALSA (mixer setup) card=%d - %s\n", card, snd_strerror(err));
+      goto errormixer;
+    }
+
+  err = snd_mixer_load(mixer);
+  if (err < 0)
+    {
+      DPRINTF(E_WARN, L_LAUDIO, "Failed to probe ALSA (mixer setup) card=%d - %s\n", card, snd_strerror(err));
+      goto errormixer;
+    }
+
+  DPRINTF(E_LOG, L_LAUDIO, "Available ALSA devices with playback, card=%d h/w name='%s' name=%s:\n", card, hwdev, snd_ctl_card_info_get_name(info));
```

Comment:
Just write one final log message instead of collating. There is no guarantee that another thread does not write a log message in between the two parts here.